nixos/modules/programs/ssh.nix

```diff
@@ -269,10 +267,10 @@ in
       [ { assertion = cfg.forwardX11 -> cfg.setXAuthLocation;
           message = "cannot enable X11 forwarding without setting XAuth location";
         }
-      ] ++ flip mapAttrsToList cfg.knownHosts (name: data: {
-        assertion = (data.publicKey == null && data.publicKeyFile != null) ||
-                    (data.publicKey != null && data.publicKeyFile == null);
-        message = "knownHost ${name} must contain either a publicKey or publicKeyFile";
```

Comment:
this use sounds like a pretty good reason to have knownHosts be an attrset right now, though that point also becomes moot if you can add the assert into to knownHosts submodule type

Reply:
Alas, you can't; I tried. If you want more specificity it could say `knownHost for [${lib.concatStringsSep ", " hostNames}]`.

Reply:
ah, just saw another use. https://github.com/NixOS/nixpkgs/pull/155517/files#diff-0acbea65dd78b9064aee4ece806fe8d84f091d3d69795ccf51fd3981f4e6a612L190

i think turning it into a list would be fine with your suggestion for the assert message, and turning `hostNames` into a nonempty list to remove reliance on the submodule name entirely (since in lists it's going to be pretty not-useful). having this default does sound nice though :confused:

Reply:
Ohhh... Now I see how it was being used. Hm.... now that I see that it doesn't seem as unreasonable to me.

 Maybe the way to do it is have `extraHostNames` instead of `hostNames`, and always include the module name in there. Anyways, I'm making the changes you suggested and we can keep discussing how is the best way to go about it.
 
 Or honestly even just changing this example:
 ```
myhost2 = {
   hostNames = [ "myhost2" ];
   publicKeyFile = ./pubkeys/myhost2_ssh_host_dsa_key.pub;
};
```
to
```
myhost2 = {
   publicKeyFile = ./pubkeys/myhost2_ssh_host_dsa_key.pub;
};
```

Reply:
an `extraHostNames` that's merged with the submodule name sounds good. doesn't break (presumably) most current uses as well, which is always a great plus :)

Reply:
I was imagining that *instead* of `hostNames`. Are you saying you think most uses don't use `hostNames`, and are more like
```
myhost2 = {
   publicKeyFile = ./pubkeys/myhost2_ssh_host_dsa_key.pub;
};
```
or were you imagining `extraHostNames` in addition to `hostNames`. That seems kind of noisy to me. I think if I encountered that I would also be confused. "which field am I supposed to set?".

One option is to add `extraHostNames` and then mark `hostNames` as deprecated.

Reply:
when I say mark `hostNames` as deprecated I mean it's still there and can be set, but produces a warning saying it's deprecated, and there will be a `mkDefault` setting `hostNames` to `[ name ] ++ extraHostNames` (just like right now it's `mkDefault`ed to `[ name ]`)

Reply:
ah! yeah, that sounds good. then eventually we can remove `hostNames`, maybe in 22.11 or so. :+1: